src/v/serde/envelope.h

```diff
@@ -67,77 +68,18 @@ struct checksum_envelope {
     static constexpr auto redpanda_serde_build_checksum = true;
 };
 
-namespace detail {
-
-template<typename T, typename = void>
-struct has_compat_attribute : std::false_type {};
-
-template<typename T>
-struct has_compat_attribute<
-  T,
-  std::void_t<decltype(std::declval<T>().redpanda_serde_compat_version)>>
-  : std::true_type {};
-
-template<typename T, typename = void>
-struct has_version_attribute : std::false_type {};
-
-template<typename T>
-struct has_version_attribute<
-  T,
-  std::void_t<decltype(std::declval<T>().redpanda_serde_version)>>
-  : std::true_type {};
-
-template<typename T, typename = void>
-struct inherits_from_envelope : std::false_type {};
-
-template<typename T>
-struct inherits_from_envelope<
-  T,
-  std::void_t<decltype(std::declval<T>().redpanda_inherits_from_envelope)>>
-  : std::true_type {};
-
+// clang-format off
 template<typename T>
-struct compat_version_has_serde_version_type {
-    static constexpr auto const value = std::is_same_v<
-      std::decay_t<decltype(std::declval<T>().redpanda_serde_compat_version)>,
-      version_t>;
+concept is_envelope = requires {
+    { T::redpanda_serde_version } -> std::convertible_to<serde::version_t>;
+    { T::redpanda_serde_compat_version } -> std::convertible_to<serde::version_t>;
```

Comment:
I'm not sure why these have dropped from `same` to `convertible`?

Reply:
thought they were simpler though, and it did not matter if they are exactly `version_t` or can be converted to `version_t`.

i'd be happy to revert to `same_as<>`, if you believe it's important.

Reply:
They're serialised without a size check or a cast, so they'd better be `uint8_t`: https://github.com/tchaikov/redpanda/blob/cc7a2b4678725f21a4b0572eb7847fae40db7aff/src/v/serde/serde.h#L144-L145

Reply:
@BenPope thank you for your explanation! i am using
```c++
std::same_as<std::remove_cvref_t<decltype(T::redpanda_serde_version)>,
              serde::version_t>
```
now. was trying to make this simpler in hope to drop the `#clang-format off` guard, with something like
```c++
template<typename T,
         typename Version = std::add_lvalue_reference_t<std::add_const_t<serde::version_t>>>
concept is_envelope = requires {
  { T::redpanda_serde_version } -> std::same_as<Version>;
  { T::redpanda_serde_compat_version } -> std::same_as<Version>;
};
```
but, that might be kind of overkill. so the current implementation is used instead.

Reply:
Yes, I meant to use the `serde::version_t` alias.

It looks ok to me, or:
```c++
template<typename T>
concept is_envelope = requires {
  { T::redpanda_serde_version } -> std::same_as<const serde::version_t&>;
  { T::redpanda_serde_compat_version } -> std::same_as<const serde::version_t&>;
};
```